fix: NetworkAnimator destinationStateMachine exception [MTT-5083]

NoelStephensUnity opened the pull request:
This resolves the issue where `NetworkAnimator` was not checking for  `AnimatorStateTtansition.destinationStateMachine`, which was causing an exception, and adding any possible states with triggers defined within it.  This also prevents the server from sending itself messages about animation state changing (#2305).

[MTT-5083](https://jira.unity3d.com/browse/MTT-5083)
[MTT-5122](https://jira.unity3d.com/browse/MTT-5122)

GitHub Issues Resolved:
#2305
#2293 
#2374
#2366
#2368

## Changelog
- Fixed: `NetworkAnimator` issue where it was not checking for `AnimatorStateTtansition.destinationStateMachine` and any possible sub-states defined within it.
- Fixed: `NetworkAnimator` issue where the host client was receiving the ClientRpc animation updates when the host was the owner.
- Fixed: `NetworkAnimator` issue with using pooled objects and when specific properties are cleaned during despawn and destroy.
- Fixed: issue where `NetworkAnimator` was checking for animation changes when the associated `NetworkObject` was not spawned.

## Testing and Documentation
- Includes integration test updates.
- Includes manual test updates.

NoelStephensUnity committed:
fix

This fixes the issue where NetworkAnimator was not taking into consideration `AnimatorStateTtansition.destinationStateMachine`.  Now, NetworkAnimator parses each destinationStateMachine's states for trigger transitions and includes those in the transition table.

NoelStephensUnity committed:
fix

passing setTrigger bool to internal method.

NoelStephensUnity committed:
test manual

Manual test that validates this fix.

NoelStephensUnity committed:
fix

Only send animation updates to clients and not host client when animation state changes.

NoelStephensUnity committed:
test

MTT-5083
Updating test to validate states defined within a destinationStateMachine are triggering.

NoelStephensUnity committed:
update

MTT-5083
Adding entry for this fix.

NoelStephensUnity committed:
update

MTT-5083
Adding the PR number to the changelog entry.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-destinationstatemachine-exception

jeffreyrainy reviewed (dismissed):
.

jeffreyrainy reviewed (commented):
.

jeffreyrainy reviewed (commented):
Mistakenly approved wrong PR, please wait for actual review before acting on this approval.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-destinationstatemachine-exception

NoelStephensUnity committed:
update and fix

This includes some updates to NetworkAnimator that will remove the need to synchronize client's using RPCs but instead use the new OnSynchronize method.

Fixed an issue with using pooled objects and when specific properties are cleaned during despawn and destroy.

NoelStephensUnity committed:
test

Made components visible to testproject again although I might try to figure out a better way to handle these two ways of testing.
Updated the NetworkAnimatorTests to acquire the state info from NetworkAnimator.SynchronizationStateInfo.

NoelStephensUnity committed:
fix

Making the NGO Components AssemblyInfo adjustment only expose internals if NGO_COMPONENTS_ASSEMBLY_RUNTIME_VISIBILITY is defined.

NoelStephensUnity committed:
test

Removed enabling verbose debug in LateJoinSynchronizationTest

NoelStephensUnity committed:
fix

Adjusting to original serialized format for backwards compatibility.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-destinationstatemachine-exception

NoelStephensUnity committed:
update

removed the AnimationSynchronizationData class as it really was not needed.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-destinationstatemachine-exception

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-destinationstatemachine-exception

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-destinationstatemachine-exception

NoelStephensUnity committed:
fix

This fixes the issue where NetworkAnimator will not check for animation changes unless the associated NetworkObject is spawned.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-destinationstatemachine-exception

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-destinationstatemachine-exception

netcode-ci-service committed:
Merge develop into fix/networkanimator-destinationstatemachine-exception

netcode-ci-service committed:
Merge develop into fix/networkanimator-destinationstatemachine-exception

NoelStephensUnity committed:
update

Removing the NGO_COMPONENTS_ASSEMBLY_RUNTIME_VISIBILITY from the scriptingDefineSymbols.
Replacing it with UNITY_INCLUDE_TESTS in the components AssemblyInfo.cs file.
Adding an additional remark about the context of when CheckForAnimatorChanges is invoked.

NoelStephensUnity committed:
update

Migrating all test related assemblies within the UNITY_INCLUDE_TESTS define region.

NoelStephensUnity committed:
update

Minor fix for weapon attack option in manual test.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-destinationstatemachine-exception

NoelStephensUnity committed:
update

updating what was fixed in this PR
